### DISCUSSION TITLE
Code Review - ParametrizedLogger struct, references class instances

### DISCUSSION THREAD

**Discussion by najak3d** [AvaloniaUI/Avalonia#7715] [Q&A] [answered]
**This post is a code review of Avalonia source code, for ParametrizedLogger:**

Over a decade ago, I learned that it's "bad to have structures reference class instances" (as it drives the GC wild, tracking these strong references around).  Maybe things have changed since I learned this convention, or maybe this particular situation warrants a violation of this rule-of-thumb.

Here's the code in question:
```c#
    public readonly struct ParametrizedLogger
    {
        private readonly ILogSink _sink;
        private readonly LogEventLevel _level;
        private readonly string _area;
```

Both _sink and _area are class instances on the heap.

For situations like this where you are wanting to create a form of Flyweight, IMO, it's best to have the struct just hold onto an "InstanceIndex" which then references statically defined Arrays of actual class references.   Thus the first instantiated ParametrizedLogger becomes Instance = 1,... which references a local static array of "ILogSink" for near instant lookup on demand.

Or if that pattern isn't acceptable, then I would think that ParametrizedLogger should become a **class**, instead of struct.

==
Am I missing something here? 

I want to always reiterate my love for Avalonia, and appreciation for the whole team who brought this awesome product to fruition.  This team and product is awesome.

**Comment by MarchingCube** [comment]
TBH I've never heard of this claim, most of recent .net work was about making structs better/faster. Only issue that I recall was that using byref structs was a bit slower due to interior pointer tracking (might be better now). But Avalonia is not doing that a lot since it still has to target netstandard 2.0 which has poor support for spans.

From my perspective statically defined arrays sound like a bigger problem due to introducing an extra GC root and potential of leaks.

**Reply by najak3d** [reply]
> TBH I've never heard of this claim, most of recent .net work was about making structs better/faster. Only issue that I recall was that using byref structs was a bit slower due to interior pointer tracking (might be better now). But Avalonia is not doing that a lot since it still has to target netstandard 2.0 which has poor support for spans.
> 
> From my perspective statically defined arrays sound like a bigger problem due to introducing an extra GC root and potential of leaks.

When it comes to situations like this, we're going to always have fewer than 10 LogSinks, right?  (maybe only 1?)   The spot in code where you create these log sinks you put some simple detection that starts to warn you if you get over a certain acceptable number -- which would indicate that people must be re-creating LogSinks by accident

I use this technique a lot for static/global contexts/instances for which everyone needs to access, but they are few in number.  Such as "Scene", "Loggers", "Viewport", "View", "Context" -- There are only 1-2 of these at any time - so I pass around "SceneRef struct" which is just the Scene Index, which implicit converts to/from the actual Scene.   Then every Scene knows it's own Index.

This keeps reference counts down, and less work for the GC.   And it's easy and low-risk.

In the case of my code review here -- I think they've chosen the worst option which is to create a struct with reference pointers.   You might as well make this Logger itself into a class, vs. a struct with two-references.   I believe each time you pass around this current struct, it's the equivalent of passing two references at once - so a double whammy.  

1. If Logger were a class -- it's like passing 1 reference
2. If Logger were a Flyweight Struct Index - it's like passing 0 References.
3. But since Logger is a Struct, it's like passing 2 references each time.

So if the authors don't want to make this a Flyweight, then it seems like they should at least make it a class, not a struct.

This is not huge deal, but this is more about establishing fact/standard -- "is this a good practice?"  And if we establish the answer as "no", then that standard can be applied throughout -- if applicable.  (but this may be the only case)

**Reply by MarchingCube** [reply]
There is generally a single log sink. Parametrized sinks encode log level and area info and it is hard to predict all combinations of it.

I did a quick benchmark to check performance and it seems using either a struct or a class is almost the same (minus class variant allocating extra garbage). There is some variance though (I had runs that showed class variant as being faster marginally).

|    Method |     Mean |    Error |   StdDev | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------- |---------:|---------:|---------:|------:|-------:|------:|------:|----------:|
| LogStruct | 10.15 ns | 0.102 ns | 0.096 ns |  1.00 |      - |     - |     - |         - |
|  LogClass | 11.93 ns | 0.046 ns | 0.036 ns |  1.18 | 0.0096 |     - |     - |      40 B |

Due to the nature of structs passing it around can be fairly cheap (due to `in`/`ref`) modifiers but even that can be a deoptimization as struct members can also end up in registers and get passed fairly fast.

But anyway - this benchmark is showing overheads in fractions of a nanosecond. And because this is actually logging after resolving such logger the app will actually write to the log which will cause 99% of time spent anyway.

Also about point 2 - it will be like passing 0.5 - 1 reference depending the on architecture (32 bit vs 64). But it will also require an extra access indirection to a global array somewhere (and a bounds check most likely) which can remove all gains. Especially with modern CPUs being really fast at executing instructions and quite slow at accessing random memory (latencies are huge even for CPU caches).

One can also see .NET runtime using structs virtually anywhere it is possible due to perf/memory gains due to better cache locality and reduced allocations. Handle (so basically a number) based approach has its own benefits (removing potential leaks, allowing for replacing target contents, centralizing management) but I don't see any immediate benefit for logger.

But if you can prepare a benchmark showing reasonable performance gains for such approach then we can definitely change this approach. Since in general about establishing practices - I think any request/change related to performance should provide benchmark info. I've tried to optimize many things in Avalonia and I've seen weirdest regressions due to small changes (for example - changing to pass math types by reference instead of by value which caused many regressions due to worse code generation).

On the other hand - there are other areas of the framework that could use more perf tuning (like ongoing work on visual tree storage or property system) and gains there can give performance gains orders of magnitude better than tweaking logger.

**Reply by najak3d** [reply]
@MarchingCube  - Turns out my knowledge of this is dated back 15 years.  I think the addition of "readonly" structs in C# 7.2 may have obliterated my GC workload concerns, well enough.

I've been under the long-standing belief that "structs should never hold references" - but apparently this isn't taboo anymore.   So when I saw this, it looked to me like an act of Taboo.

**So I'll retract my whole code review here.    I'm guessing that as of C# 7.2, this taboo is no longer taboo.**

**Comment by najak3d** [comment, marked as answer]
**I retract this criticism.   This code looks fine.**

From what I can tell this act of putting adding reference types to a struct, is no longer Taboo, perhaps made acceptable via "readonly structs" starting with C# 7.2.

When I saw a "struct with references", it triggered my long-standing belief that "this is NEVER a good idea" -- but this belief may now be outdated and now moot/incorrect.     But I've changed my view; this usage of a struct seems to be just fine.
